numpy/core/src/multiarray/dtypemeta.h

```diff
@@ -1,7 +1,65 @@
 #ifndef _NPY_DTYPEMETA_H
 #define _NPY_DTYPEMETA_H
 
+
+/* DType flags, currently private, since we may just expose functions */
+#define NPY_DT_LEGACY 1 << 0
+#define NPY_DT_ABSTRACT 1 << 1
+#define NPY_DT_PARAMETRIC 1 << 2
+
+
+typedef struct {
+    /* DType methods, these could be moved into its own struct */
+    discover_descr_from_pyobject_function *discover_descr_from_pyobject;
+    is_known_scalar_type_function *is_known_scalar_type;
+    default_descr_function *default_descr;
+    common_dtype_function *common_dtype;
+    common_instance_function *common_instance;
+    /*
+     * The casting implementation (ArrayMethod) to convert between two
+     * instances of this DType, stored explicitly for fast access:
+     */
+    PyObject *within_dtype_castingimpl;
+    /*
+     * Dictionary of ArrayMethods representing most possible casts
+     * (structured and object are exceptions).
+     * This should potentially become a weak mapping in the future.
+     */
+    PyObject *castingimpls;
+
+    /*
+     * Storage for `descr->f`, since we may need to allow some customizatoin
+     * here at least in a transition period and we need to set it on every
+     * dtype instance for backward compatibility.  (Keep this at end)
+     */
+    PyArray_ArrFuncs f;
+} NPY_DType_Slots;
+
+
 #define NPY_DTYPE(descr) ((PyArray_DTypeMeta *)Py_TYPE(descr))
+#define NPY_DT_SLOTS(dtype) ((NPY_DType_Slots *)(dtype)->dt_slots)
+
+#define NPY_DT_is_legacy(dtype) ((dtype)->flags & NPY_DT_LEGACY)
+#define NPY_DT_is_abstract(dtype) ((dtype)->flags & NPY_DT_ABSTRACT)
+#define NPY_DT_is_parametric(dtype) ((dtype)->flags & NPY_DT_PARAMETRIC)
+
+/*
+ * Macros for convenient classmethod calls, since these require
+ * the DType both for the slot lookup and as first arguments.
+ *
+ * (Macros may include NULL checks where appropriate)
+ */
+#define NPY_DT_call_discover_descr_from_pyobject(dtype, obj)  \
```

Comment:
I'd maybe capitalize as `NPY_DT_CALL_discover_descr_from_pyobject` to make the commonality of these macros even clearer. Or `NPY_DT_METH_discover...`, etc. (or `C_METH`?)

Reply:
Good idea, will go with the CALL for now.